### PR TITLE
Use slotted dataclasses instead of namedtuples.

### PR DESCRIPTION
**Pull request by JustinTArthur**
This provides much better performance over plain namedtuple on cpython for the things we're constructing and transporting a lot of: txes, inputs, and outputs. It might be a little more readable.

This is inspired by earlier experiments by @ofek and benchmarks by @erasmospunk in kyuupichan/electrumx#150.

### CODE REVIEW

**Comment by JustinTArthur** [comment]
Just need to fix the unit tests and make sure we're not comparing these to tuple hashes anywhere before this is ready.

**Comment by SomberNight** [comment]
Looks good. I like type annotations :P

File: electrumx/lib/coins.py [hunk 1]
```diff
@@ -51,7 +51,13 @@
                                       SmartCashElectrumX, AuxPoWElectrumX)
 
 
-Block = namedtuple("Block", "raw header transactions")
+@dataclass
+class Block:
+    __slots__ = "raw", "header", "transactions"
+    raw: ByteString
+    header: ByteString
```
Comment (SomberNight) [inline electrumx/lib/coins.py:58]: What are the actual types used here? Is these not just `bytes`?
https://github.com/spesmilo/electrumx/blob/7e1a0288537c1cf1d202fbadff3a7f0f93b7565e/electrumx/server/daemon.py#L229
I've never used use `ByteString` to annotate before, just looking at https://github.com/python/mypy/issues/4871
but if you think it fits; fine by me.

Comment (JustinTArthur) [reply electrumx/lib/coins.py]: It was a placeholder until I could confirm we weren't constructing a bytearray.

**Comment by JustinTArthur** [comment]
I tried using `frozen=True` to give the immutability we had before w/ namedtuple, but it wiped most of the performance boost.

Now that they aren't frozen, it means we can't use them as-is in `set`s or as keys for `dict`s. I'm ok with this.

**Comment by ecdsa** [comment]
these changes look good to me. 
(shitcoins hurt my eyes, though)

**Pull request merged**